components/tenant-fetcher/cmd/main.go

```diff
@@ -0,0 +1,201 @@
+	logger := log.C(ctx)
+
+	handlerWithTimeout, err := timeouthandler.WithTimeout(handler, timeout)
+	exitOnError(err, "Error while configuring tenant mapping handler")
+
+	srv := &http.Server{
+		Addr:              address,
+		Handler:           handlerWithTimeout,
+		ReadHeaderTimeout: timeout,
+	}
+
+	runFn := func() {
+		logger.Infof("Running %s server on %s...", name, address)
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			logger.Errorf("%s HTTP server ListenAndServe: %v", name, err)
+		}
+	}
+
+	shutdownFn := func() {
+		logger.Infof("Shutting down %s server...", name)
+		if err := srv.Shutdown(context.Background()); err != nil {
+			logger.Errorf("%s HTTP server Shutdown: %v", name, err)
+		}
+	}
+
+	return runFn, shutdownFn
+}
+
+func getOnboardingHandlerFunc(tenantPathParam string) func(writer http.ResponseWriter, request *http.Request) {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		logger := log.C(request.Context())
+
+		logHandlerRequest("onboarding", tenantPathParam, request)
+		if err := logBody(request, writer); err != nil {
+			logger.Error(errors.Wrapf(err, "while logging request body"))
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		writer.Header().Set("Content-Type", "text/plain")
+		writer.WriteHeader(http.StatusOK)
+		if _, err := writer.Write([]byte(compassURL)); err != nil {
+			logger.Error(errors.Wrapf(err, "while writing response body"))
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func getDecommissioningHandlerFunc(tenantPathParam string) func(writer http.ResponseWriter, request *http.Request) {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		logger := log.C(request.Context())
+
+		logHandlerRequest("decommissioning", tenantPathParam, request)
+		if err := logBody(request, writer); err != nil {
+			logger.Error(errors.Wrapf(err, "while logging request body"))
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		writer.Header().Set("Content-Type", "application/json")
+		err := json.NewEncoder(writer).Encode(map[string]interface{}{})
+		if err != nil {
+			logger.Error(errors.Wrapf(err, "while writing to response body"))
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func logHandlerRequest(operation, tenantPathParam string, request *http.Request) {
+	tenantID := mux.Vars(request)[tenantPathParam]
+	log.C(request.Context()).Infof("Performing %s for tenant with id %q", operation, tenantID)
+}
+
+func logBody(r *http.Request, w http.ResponseWriter) error {
+	logger := log.C(r.Context())
+
+	buf, bodyErr := ioutil.ReadAll(r.Body)
+	if bodyErr != nil {
+		logger.Info("Body Error: ", bodyErr.Error())
+		http.Error(w, bodyErr.Error(), http.StatusInternalServerError)
+		return nil
+	}
+
+	rdr1 := ioutil.NopCloser(bytes.NewBuffer(buf))
+	rdr2 := ioutil.NopCloser(bytes.NewBuffer(buf))
+	logger.Infof("Body: %q", rdr1)
+	r.Body = rdr2
```

Comment:
Why the need for both readers? In the log message you can just use the `buf` variable from line 182 to print the body.

Reply:
one was for setting a new body since the first one was read (not that anyone else would try read anyway..), the second was for better formatting of the body (tbh i dont remember if it was better or the same)

Reply:
I think that the first one `rdr1` is not needed as it just creates another reader out of the same buffer.